### ISSUE TITLE
TTable::sync in a ko situation

### ISSUE THREAD

**Issue by mkw1234** [gcp/leela-zero#576]
Following two positions (case 1 and case 2) have equal hash values but do not have same positional values.  However, Leelaz MCTS might treat these two positions as equal when a node is synced with another node in the TTable of equal hash value.  Shouldn't the nodes not the be synced in this case?

![image](https://user-images.githubusercontent.com/35014179/34636960-68ceb2a4-f2ef-11e7-9636-f13e72cfb221.png)
**Case 1**

![image](https://user-images.githubusercontent.com/35014179/34636966-74076652-f2ef-11e7-9199-b330c92db8f5.png)
**Case 2**

**Comment by gcp** [comment]
Shouldn't the prisoners by unequal if it is a ko?

**Comment by mkw1234** [comment]
No, not necessarily.  In two positions shown above there is 1 black stone prisoner.  Same hash values.  The difference is in case 1, black can take the ko, but in case 2 black can not take the ko.

**Comment by ThePokerDude** [comment]
>  Shouldn't the prisoners be unequal if it is a ko? They are included in the hash. <

Not necessarily. For examlpe: 
1) (move number for example 24 (e6f6 already played))  last move = e4 wins opponent's stone -> ko 
2) (move number 22 - e6f6 not played yet) last moves: e4 wins opponent's stone, e6 f6 -> no ko

**Comment by joseph-ireland** [comment]
This can happen in a triple ko too. The only fully correct solution I can see is to enumerate the superko-illegal points for each position in advance and include them in the hash.

**Comment by mkw1234** [comment]
In addition to ko issue, I think there might be a general problem with the syncing concept.  In baduk,  correct timing of moves is very important.  If the program plays a good sequence of moves to reach a good position, NN should be rewarded.  If the program plays bad sequence of moves but reaches same position by accident, NN should not be rewarded for that.  But by syncing a node to another node based on matching board configuration, NN might be rewarded for a bad sequence of moves and develop a bad habit.

**Comment by gcp** [comment]
Your last comment makes no sense. The evaluation fed back to the root is the one from the leaves. If it comes from a bad sequence, that means the opponent could have forced an inferior branch earlier up the tree and have stopped you from reaching it.

**Comment by gcp** [comment]
>The difference is in case 1, black can take the e4 stone, but in case 2 black can not take the e4 stone.

Ah, that makes sense. The issue is the path dependence. Including the ko move, if any, in the hash should solve the most common cases and may be worthwhile. The general problem is known and happens in chess-programs too with threefold repetitions.

**Comment by mkw1234** [comment]
I have the stated doubts about Transposition Tables being appropriate for MCTS guided by NN.   What you really have here is MCTS guided by a hybrid of NN + Transposition Table.

**Comment by alreadydone** [comment]
So we've been syncing two nodes whose previous seven positions don't agree? Such nodes don't even have the same NN evaluations! Or do you think (and did tests showing) that the difference would be negligible if no ko was involved?

I  think @mkw1234 is at least right in that the history affects the NN evaluation at least in AGZ.

**Comment by gcp** [comment]
>So we've been syncing two nodes whose previous seven positions don't agree

If the positions are the same, yes. Note that the evaluation should be the same, aside from the issue with ko highlighted here (which should be fixed by treating them as different). If the NN gives a different evaluation for the exact same board position (which it can certainly do), that's just because it has not reached the full truth about Go yet. There is no reason to not reuse an evaluation that is as correct as another.

>help the NN learn the timing.

This is a meaningless statement.

**Comment by billyswong** [comment]
https://senseis.xmp.net/?Cycle

I have no idea how do we know which 2 game states are "ko-equivalent" or not without pouring in a lot of game knowledge, when there are so many kinds and lengths of cycle.

**Comment by gcp** [comment]
FWIW after the addition of the NNCache, and knowing about this ko bug, it seems worthwhile to me to check if the TTable actually makes the program stronger, and I'm testing this now.

**Comment by gcp** [comment]
>I have no idea how do we know which 2 game states are "ko-equivalent" or not without pouring in a lot of game knowledge, when there are so many kinds and lengths of cycle.

It's not necessary to handle all cases correctly, just behave decently on the common ones. 

Many heuristics are useful even if they are wrong sometimes or have other downsides. So pointing out a heuristics can go wrong is not useful or constructive unless you a) have a better replacement b) can demonstrate the heuristic makes things worse.

**Comment by joseph-ireland** [comment]
@billyswong for checking cycles/triple ko, leela-zero keeps a list of "ko hashes" (seen above, unique for each board position) for prior positions, and before searching a suggested move, it checks if its ko hash is in the list (i.e. the position has been seen before). If it is, that branch is invalidated and it isnt considered. 

The problem here is that in specific circumstances, instead of being invalidated, the search can be merged with another search where the ko has played out differently. This is done by comparing the "hash" of two positions, which is supposed to be unique for identical situations, but currently doesnt include ko state at all. 

This isn't easy to avoid for longer cycles, since the trees could be merged early in the cycle, before any of the later moves causing a cycle are even evaluated. The problem actually affecting the game outcome would be reasonably rare anyway, even without an exotic cycle, so gcp is considering only fixing it for "regular" ko, which is easy to calculate.

The merging (transposition tables) is important for things like trades/capture races where there are many move options which are equivalent. Not using it will thin the search exponentially (based on the capture race length) and make it evaluate these trades very inaccurately. Things like timing of exchanges shouldnt be affected, since it will consider any refutations of timing errors before the trees are merged.

**Comment by grolich** [comment]
> I have the stated doubts about Transposition Tables being appropriate for MCTS guided by NN. What you really have here is MCTS guided by a hybrid of NN + Transposition Table.

Apologies in advance if I misunderstood your reasoning, but it seems to me you are assuming something about the NN positional evaluation:

You seem to make the (hidden?) assumption that the history of the last few positions/moves being fed to the NN is there to be used for somehow training better move sequences (in addition to "normal" better evaluation/move training), or that it can be used for that purpose.

Well, there are 2 parts to the answer:
1) it isn't there for that purpose at all. That can be simply stated as DeepMind, in their AGZ paper, said it's simply there because kos aren't observable by the NN otherwise.

2) It can't be used for such a purpose the way you seem to be suggesting (not sure how the "sequence training" you mentioned would work, as gcp stated, it's not really making much sense in the context of a search like MCTS). 

The reason:

Your idea seems to be that the NN should output different results if the input planes representing the history are different for the same position (if the NN output would be exactly the same, the exact same prior probabilities would be used, and the same MCTS would occur. In that case, the TT just saves time).
However, the entire search is guided by the assumptions that the policy returns probabilities of moves being the best, and the value returns the chance of winning. 
Same positions (if there is no ko issue) have the exact same good and bad moves. The network trains to reach as close as it can to these values.
Any change to those values in a way which might make the outputs deliberately different would violate these assumptions.

In reality. the values will often be different simply because the net hasn't learned they should be the same yet, but training is geared towards making the net recognize them as the same, among other things.


The history input planes just weren't meant to be used for your suggested "sequence training", and the entire algorithm AGZ uses wasn't built on such ideas. In fact they are more likely to hurt than to help for the reasons stated above.

You would have to think about something quite different to make such an idea realistic (if at all possible), but in any case, whether it's possible or not, LZ is following the AGZ experiment as closely as it can, and I don't think it would add any big new ideas anytime soon.

**Comment by gcp** [comment]
Result of a little test:

```
leelaz-fca0 v leelaz-fca0-ntt (200/1000 games)
board size: 19   komi: 7.5
                  wins              black         white        avg cpu
leelaz-fca0        107 53.50%       52 52.00%     55  55.00%    323.08
leelaz-fca0-ntt     93 46.50%       45 45.00%     48  48.00%    301.23
                                    97 48.50%     103 51.50%
```

It seems even with a potential ko issue the transposition table does still help a bit. (The above isn't statistically significant, but it's also very unlikely to make things worse).

**Comment by mkw1234** [comment]
> Your idea seems to be that the NN should output different results if the input planes representing the history are different for the same position 

Yes.  I thought this might be beneficial to the process of NN training.

Suppose a program considers branch A first and then branch B second.

The branches take different paths but end up in same position. 

When A is evaluated at this position, NN outputs a value of 0.9.
When B is evaluated at this position, NN outputs a value of 0.5.

The program will be discouraged from taking the branch B because of the lower value. 
This is the desired behavior.

But in case of LZ, it syncs the position B to position A, which encourages branch B.
If branch B is taken, and the game won by White, the NN will be trained to prefer branch B (high evaluations for white moves 1 and 3).   Now we have a burden of unlearning these moves.

There are many many possible combinations of moves that LZ can consider that lead to the same position, but only one  is the correct one.

![image](https://user-images.githubusercontent.com/35014179/34651422-70ef6db4-f413-11e7-93d9-830ddba956ec.png)
**(A) Correct sequence**

![image](https://user-images.githubusercontent.com/35014179/34651381-ed786bc0-f412-11e7-9f15-d9bd15c67f85.png)
**(B) Incorrect sequence**

**Comment by gcp** [comment]
@mkw1234 Your reasoning really makes no sense to me. Why wouldn't we want to evaluate the end positions as identical (barring ko issues!) if they are the same position. That would be very much the goal of the training!

>If branch B is taken, and the game won by White

If branch B is taken and the game won, the moves are good and we *want* to reinforce them. It is the opponent that made a mistake by not punishing the incorrect sequence into another ending position.

**Comment by mkw1234** [comment]
> Why wouldn't we want to evaluate the end positions as identical (barring ko issues!) 

NN will eventually evaluate the end positions as identical as it become better through training.  But before then, syncing will more likely than not burden the network with wrong ideas and slow its progress.

**Comment by gcp** [comment]
>will more likely than not burden the network with wrong ideas and slow its progress.

But why would that be?

**Comment by earthengine** [comment]
> Not necessarily. For examlpe:
> 
> (move number for example 24 (e6f6 already played)) last move = e4 wins opponent's stone -> ko
> (move number 22 - e6f6 not played yet) last moves: e4 wins opponent's stone, e6 f6 -> no ko

"ko" or "no ko" only determines what is the legal move in the future. If two position have the same ko-hash (or I would say the same position with two different history), it means just that from any future position a move cannot go back to any of these positions. But obviously there is only one position you can go back to, and even you went back, you are in a position that have a different history. So why the actual history of the position you went back matter?

In other words, with history + current position we determine weather a move is legal or not. But when we do so, we only need the positions (hash) in the history. We don't care how a historical position being reached. Is it clear?

This is why I believe this is hardly an issue.

**Comment by ThePokerDude** [comment]
> In other words, with history + current position we determine weather a move is legal or not. But when we do so, we only need the positions (hash) in the history.
I agree

Maybe I misunderstood something. The point was the following:
When LeelaZ first calculates the the winning probability of a position the history matters (taking the stone might be the best move, which dependend on the history can be legal or illegal).
Let's assume that in the one case the winning probability is 0.2 and in the other 0.6.
Let's assume you save 0.2 for the position in the TTable.
When you look up the number later, depending on the history, it might be completely wrong.

**Comment by gcp** [comment]
Right, the legal moves in a position - and hence, the value of it - vary depending on whether ko moves are possible or not, and this depends on the positions' history.

This, the hash should at the very least include basic ko information. (Compare this to hashing en-passant squares in chess engines - the issue is near identical)

**Comment by gcp** [comment]
We catch and handle the basic cases of this now correctly.
